File: gw/gh_ops.py
"""GitHub CLI operations."""

import json
import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass
class PullRequestInfo:
    """Information about a pull request."""

    number: int
    state: str
    base: str | None
    url: str | None


@dataclass
class ChecksInfo:
    """Information about PR checks."""

    passed: int
    total: int
    state: str | None  # "ok", "fail", "pend", or None


def get_pr_info(repo_root: Path, branch: str) -> PullRequestInfo | None:
    """Get pull request information for a branch."""
    try:
        result = subprocess.run(
            [
                "gh",
                "pr",
                "list",
                "--state",
                "all",
                "--head",
                branch,
                "--json",
                "number,state,baseRefName,mergedAt,url",
                "--limit",
                "1",
            ],
            cwd=repo_root,
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        pr_list = json.loads(result.stdout.strip() or "[]")
    except (subprocess.CalledProcessError, json.JSONDecodeError):
        return None

    if not pr_list:
        return None

    pr = pr_list[0]
    merged_at = pr.get("mergedAt")
    state = "MERGED" if merged_at else pr.get("state", "OPEN")

    return PullRequestInfo(
        number=pr.get("number"),
        state=state,
        base=pr.get("baseRefName"),
        url=pr.get("url"),
    )


def get_checks_info(repo_root: Path, pr_number: int) -> ChecksInfo | None:
    """Get check status information for a pull request."""
    try:
        result = subprocess.run(
            ["gh", "pr", "view", str(pr_number), "--json", "statusCheckRollup"],
            cwd=repo_root,
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        checks_json = json.loads(result.stdout.strip() or "{}")
        rollup = checks_json.get("statusCheckRollup") or []
    except (subprocess.CalledProcessError, json.JSONDecodeError):
        return None

    conclusions = [item.get("conclusion") for item in rollup]
    states = [item.get("state") for item in rollup]

    return classify_checks(conclusions, states)


def classify_checks(conclusions: list[str | None], states: list[str | None]) -> ChecksInfo:
    """Classify check results into passed/total/state."""
    total = len(conclusions)
    passed = 0
    failed = False
    pending = False

    for conclusion, state in zip(conclusions, states):
        if state and state != "COMPLETED":
            pending = True
        if conclusion is None:
            pending = True
            continue
        if conclusion == "SUCCESS":
            passed += 1
        elif conclusion in {"NEUTRAL", "SKIPPED"}:
            passed += 1
        else:
            failed = True

    status: str | None = None
    if total == 0:
        status = None
    elif failed:
        status = "fail"
    elif pending:
        status = "pend"
    else:
        status = "ok"

    return ChecksInfo(passed=passed, total=total, state=status)

File: src/gh_ops.rs
use crate::models::{ChecksInfo, PullRequestInfo};
use serde::Deserialize;
use serde_json::Value;
use std::path::Path;
use std::process::Command;

#[derive(Debug, Deserialize)]
struct PrListItem {
    number: i64,
    state: Option<String>,
    #[serde(rename = "baseRefName")]
    base_ref_name: Option<String>,
    #[serde(rename = "mergedAt")]
    merged_at: Option<String>,
    url: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ChecksView {
    #[serde(rename = "statusCheckRollup")]
    status_check_rollup: Option<Vec<Value>>,
}

fn run_gh(args: &[&str], repo_root: &Path) -> Option<String> {
    let output = Command::new("gh")
        .args(args)
        .current_dir(repo_root)
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    Some(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

pub fn get_pr_info(repo_root: &Path, branch: &str) -> Option<PullRequestInfo> {
    let stdout = run_gh(
        &[
            "pr",
            "list",
            "--state",
            "all",
            "--head",
            branch,
            "--json",
            "number,state,baseRefName,mergedAt,url",
            "--limit",
            "1",
        ],
        repo_root,
    )?;

    let list: Vec<PrListItem> = serde_json::from_str(&stdout).ok()?;
    let first = list.into_iter().next()?;

    let state = if first.merged_at.is_some() {
        "MERGED".to_string()
    } else {
        first.state.unwrap_or_else(|| "OPEN".to_string())
    };

    Some(PullRequestInfo {
        number: first.number,
        state,
        base: first.base_ref_name,
        url: first.url,
    })
}

pub fn get_checks_info(repo_root: &Path, pr_number: i64) -> Option<ChecksInfo> {
    let stdout = run_gh(
        &[
            "pr",
            "view",
            &pr_number.to_string(),
            "--json",
            "statusCheckRollup",
        ],
        repo_root,
    )?;

    let parsed: ChecksView = serde_json::from_str(&stdout).ok()?;
    let rollup = parsed.status_check_rollup.unwrap_or_default();

    let conclusions: Vec<Option<String>> = rollup
        .iter()
        .map(|item| {
            item.get("conclusion")
                .and_then(Value::as_str)
                .map(ToOwned::to_owned)
        })
        .collect();
    let states: Vec<Option<String>> = rollup
        .iter()
        .map(|item| {
            item.get("state")
                .and_then(Value::as_str)
                .map(ToOwned::to_owned)
        })
        .collect();

    Some(classify_checks(&conclusions, &states))
}

pub fn classify_checks(conclusions: &[Option<String>], states: &[Option<String>]) -> ChecksInfo {
    let total = conclusions.len() as i64;
    let mut passed = 0_i64;
    let mut failed = false;
    let mut pending = false;

    for (conclusion, state) in conclusions.iter().zip(states.iter()) {
        if state.as_deref().is_some_and(|v| v != "COMPLETED") {
            pending = true;
        }

        match conclusion.as_deref() {
            Some("SUCCESS") | Some("NEUTRAL") | Some("SKIPPED") => passed += 1,
            Some(_) => failed = true,
            None => pending = true,
        }
    }

    let status = if total == 0 {
        None
    } else if failed {
        Some("fail".to_string())
    } else if pending {
        Some("pend".to_string())
    } else {
        Some("ok".to_string())
    };

    ChecksInfo {
        passed,
        total,
        state: status,
    }
}
